packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp

```diff
@@ -128,6 +128,130 @@ void Add(
+    const KCRS& A,
+    const ISC scalarA,
+    const KCRS& B,
+    const ISC scalarB,
+    const local_map_type& AcolMap,
+    const local_map_type& BcolMap,
+    GlobalOrdinal minGlobalCol,
+    GlobalOrdinal globalNumCols,
+    Teuchos::RCP<ValuesArray>& Cvals,
+    Teuchos::RCP<RowPtrsArray>& Crowptrs,
+    Teuchos::RCP<GlobalColIndsArray>& Ccolinds);
+
+  /// \brief Given two matrices in CRS format, return their sum
+  /// \pre A and B don't need to be sorted
+  /// \param Avals Values array for A
+  /// \param Arowptrs Row pointers array for A
+  /// \param Acolinds Column indices array for A
+  /// \param scalarA Scaling factor for A
+  /// \param Bvals Values array for B
+  /// \param Browptrs Row pointers array for B
+  /// \param Bcolinds Column indices array for B
+  /// \param scalarB Scaling factor for B
+  /// \param globalNumCols The global size of the column map
+  /// \param[Out] Cvals Values array for C (allocated inside function)
+  /// \param[Out] Crowptrs Row pointers array for C (allocated inside function)
+  /// \param[Out] Ccolinds Column indices array for C (allocated inside function)
+  static void Unsorted(
+    const ValuesArray& Avals,
+    const RowPtrsArray& Arowptrs,
+    const ColIndsArray& Acolinds, 
+    const ISC scalarA,
+    const ValuesArray& Bvals,
+    const RowPtrsArray& Browptrs,
+    const ColIndsArray& Bcolinds, 
+    const ISC scalarB,
+    GlobalOrdinal numGlobalCols,
+    Teuchos::RCP<ValuesArray>& Cvals,
+    Teuchos::RCP<RowPtrsArray>& Crowptrs,
+    Teuchos::RCP<ColIndsArray>& Ccolinds);
+
+  /// \brief Given two matrices in CRS format, return their sum
+  /// \pre A and B don't need to be sorted
+  /// \param Avals Values array for A
+  /// \param Arowptrs Row pointers array for A
+  /// \param Acolinds Column indices array for A
+  /// \param scalarA Scaling factor for A
+  /// \param Bvals Values array for B
+  /// \param Browptrs Row pointers array for B
+  /// \param Bcolinds Column indices array for B
+  /// \param scalarB Scaling factor for B
+  /// \param globalNumCols The global size of the column map
+  /// \param[Out] Cvals Values array for C (allocated inside function)
+  /// \param[Out] Crowptrs Row pointers array for C (allocated inside function)
+  /// \param[Out] Ccolinds Column indices array for C (allocated inside function)
+  static void UnsortedGlobalCols(
+    const ValuesArray& Avals,
+    const RowPtrsArray& Arowptrs,
+    const GlobalColIndsArray& Acolinds, 
+    const ISC scalarA,
+    const ValuesArray& Bvals,
+    const RowPtrsArray& Browptrs,
+    const GlobalColIndsArray& Bcolinds, 
+    const ISC scalarB,
+    GlobalOrdinal numGlobalCols,
+    Teuchos::RCP<ValuesArray>& Cvals,
+    Teuchos::RCP<RowPtrsArray>& Crowptrs,
+    Teuchos::RCP<GlobalColIndsArray>& Ccolinds);
+
+  static Teuchos::RCP<map_type> getOptimizedColMap(GlobalOrdinal ncols, const GlobalColIndsArray& gids, ColIndsArray& lids, const Teuchos::RCP<const Teuchos::Comm<int>>& comm);
```

Comment:
I meant "optimized" in the sense of owning the minimum set of global cols to cover the given entries. I could just call it makeColMapAndConvertGids since that's really what it does (and it only gets called in one place so it's OK if the name is long).